### PR TITLE
Wait for namespace to be deleted before exiting AfterEach

### PR DESCRIPTION
**Pull request by gnufied**
Ginkgo has a weird bug that - AfterEach does not get called when
testsuite exits with certain kind of interrupt (Ctrl-C for example).
More info - https://github.com/onsi/ginkgo/issues/222

We workaround this issue in Kubernetes by adding a special hook into
AfterSuite call, but AfterSuite can not be used to peforms certain
kind of cleanup because it can race with AfterEach hook and
framework.AfterEach hook will set framework.ClientSet to nil. 
Also this makes various CleanupActions order dependent. 

This presents a problem in cleaning up CSI driver and testpods. This
PR removes cleanup of driver manifest via CleanupAction because that
is not safe and racy (such as f.ClientSet may disappear!). But if we
wait for Namespace to be deleted in AfterSuite hook (same hook runs
via framework.AfterEach as well) - we can give time to the testsuite
to clean itself cleanly.

/sig storage

cc @jsafrane @smarterclayton

### CODE REVIEW

File: test/e2e/storage/utils/create.go [hunk 1]
```diff
@@ -141,14 +141,7 @@ func PatchItems(f *framework.Framework, items ...interface{}) error {
 // - only the latest stable API version for each item is supported
 func CreateItems(f *framework.Framework, items ...interface{}) (func(), error) {
 	var destructors []func() error
-	var cleanupHandle framework.CleanupActionHandle
```
Comment (smarterclayton) [inline test/e2e/storage/utils/create.go:144]: I don't understand why this change was made - what's the difference here?

File: test/e2e/framework/framework.go [hunk 2]
```diff
@@ -393,7 +394,14 @@ func (f *Framework) AfterEach() {
 					} else {
 						Logf("Namespace %v was already deleted", ns.Name)
 					}
+				} else {
```
Comment (smarterclayton) [inline test/e2e/framework/framework.go:397]: So... why not just delete the namespaces yourself in a defer in your test?  Defer should still get a chance to run on ctrl+c

Comment (smarterclayton) [reply test/e2e/framework/framework.go:397]: If ctrl+c isn't allowing defers in the main body of a test to run we have a bigger problem.

File: test/e2e/storage/utils/create.go (hunk 1 above)
Comment (gnufied) [reply test/e2e/storage/utils/create.go:144]: The reason this change was made is because - cleanupActions are stored in a global map and this hook and hook installed in `framework.go` (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go#L186) might run out of order and what will happen is - if hook in framework.go ran before this hook then it will set `f.ClientSet` to nil - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go#L409 which will prevent this hook from working.

I am also thinking of making cleanupActions run in order (i.e last hook runs first)  but @thockin has tried in past to prevent hooks from being dependent on each other. 

There is one more reason why I have removed this code - This code here tears down CSI driver before testpod has been deleted. This obviously prevents testpod from cleaning up. We should rather let cleanup action delete the testpod and csidriver in order.

File: test/e2e/framework/framework.go (hunk 2 above)
Comment (gnufied) [reply test/e2e/framework/framework.go:397]: When ginkgo runner receives `ctrl-c` - it starts executing `afterSuite` - https://github.com/onsi/ginkgo/blob/master/internal/specrunner/spec_runner.go#L237 and exits immediately. If we don't cause process to wait in `afterSuite` (i.e what waiting for namespace to be deleted causes) then e2e process will exit without running defer or `aftereach`.

Comment (gnufied) [reply test/e2e/framework/framework.go:397]: `ctrl-c` does not allow defers in the main body of test to run.  

```
C
---------------------------------------------------------
Received interrupt.  Running AfterSuite...
^C again to terminate immediately
Apr  8 11:23:24.313: INFO: Running AfterSuite actions on all nodes
Apr  8 11:23:24.313: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "csi-mock-volumes-9650" for this suite.
Apr  8 11:23:24.317: INFO: Running AfterSuite actions on node 1
{"msg":"Test Suite completed","total":6,"completed":0,"skipped":486,"failed":0}

Ran 1 of 487 Specs in 8.019 seconds
FAIL! -- 1 Passed | 0 Failed | 0 Pending | 486 Skipped
```

As you can see cleanup function mentioned in defer() ^ did not run.

Comment (thockin) [reply test/e2e/framework/framework.go:397]: I tried to solve this a long time ago and failed.  Ginkgo needs to be cut out like cancer.

File: test/e2e/storage/utils/create.go (hunk 1 above)
Comment (thockin) [reply test/e2e/storage/utils/create.go:144]: FWIW - I have lost almost all context on how this works :)

**Comment by thockin** [comment]
@BenTheElder @Katharine @mithrav - any bandwidth or context to help review this?

File: test/e2e/framework/framework.go (hunk 2 above)
Comment (jsafrane) [reply test/e2e/framework/framework.go:397]: I've been playing with this a bit. SIGINT is handled by ginkgo, it does not call any defers or anything. A  test continue doing whatever it was doing before SIGINT.

In a separate goroutine, ginkgo runs AfterSuite, that deletes all namespaces, sets `f.ClientSet` to nil and call `os.Exit`. All this while the test is still testing! `os.Exit` won't call any defers either.

With this PR, AfterSuite waits a bit for the namespaces to be deleted. During this time, the test has at least chance to notice that something's wrong (e.g. it can't create new pods) and "fail" in a natural way (`framework.ExpectNoError`), which calls defers and `AfterEach`.

Still, the test namespaces are Terminating, which means Kubernetes will kill everything there in a random order. This is not exactly what the test needs - it may want to delete some stuff earlier than other, that's why we have `defers` and `AfterEach`. At least CSI tests suffers from this, CSI driver pod(s) must be deleted *after* a pod and PVC that uses a volume provided by the CSI driver. Kubelet can't unmount a volume, detach it or delete it without the driver still running.

Even with this drawback, I think it will help cleaning up some tests.

**Comment by BenTheElder** [comment]
I'm pretty over capacity as usual... :grimacing: I don't think @mithrav is active on github.

**Comment by BenTheElder** [comment]
cc @SataQiu @neolit123 for [test/e2e/framework](https://github.com/kubernetes/kubernetes/blob/548124828c948d69945cc3ef93a6136228bc3d1e/test/e2e/framework/OWNERS#L19-L20)

File: test/e2e/framework/framework.go (hunk 2 above)
Comment (neolit123) [reply test/e2e/framework/framework.go:397]: @jsafrane

> With this PR, AfterSuite waits a bit for the namespaces to be deleted. During this time, the test has at least chance to notice that something's wrong (e.g. it can't create new pods) and "fail" in a natural way (framework.ExpectNoError), which calls defers and AfterEach.

IMO having WaitForNamespacesDeleted here is fine.
related https://github.com/kubernetes/kubernetes/issues/85541

just noting that for a ctrl+c on a finalizer test we might face https://github.com/kubernetes/kubernetes/issues/70916
so having the timeout of 5 minutes (as is in this PR) is a good idea.

> In a separate goroutine, ginkgo runs AfterSuite, that deletes all namespaces, sets f.ClientSet to nil and call os.Exit. All this while the test is still testing! os.Exit won't call any defers either.

that particular os.Exit call in ginkgo is bad, so we have to think of ways of avoiding it while we still need work to get done.

@ gnufied 
> Ginkgo has a weird bug that - AfterEach does not get called when
testsuite exits with certain kind of interrupt (Ctrl-C for example).
More info - onsi/ginkgo#222

my general question here would be why do we need to care about interupts outside of local / debugging scenarios?

cc @pohly @oomichi

**Comment by gnufied** [comment]
opened another variant of this fix - https://github.com/kubernetes/kubernetes/pull/90773

**Comment by jsafrane** [comment]
can this be closed now that #90773 is merged?

**Comment by gnufied** [comment]
Yeah. Closing this for now, since we have kinda worked around this issue in - https://github.com/kubernetes/kubernetes/pull/90773
